Added support for Confluence Cloud

awanlin opened the pull request:
## Hey, I just made a Pull Request!

Added support for Confluence Cloud to the `confluence:transform:markdown` action:

- Updated URL parsing
- Updated token usage
- Added new config setting

#### :heavy_check_mark: Checklist

<!--- Please include the following in your Pull Request when applicable: -->

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [x] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/helpers.test.ts:
```diff
@@ -0,0 +1,40 @@
+
+describe('createConfluenceVariables', () => {
+  it('should return values for Confluence Url', async () => {
+    const url = 'https://confluence.example.com/display/SPACEKEY/Page+Title';
```
So there's two tests. This one tests the current implementation which from the sounds of it won't work for all Confluence on-prem instances. 

The way I made my changes the existing flow would still work but clearly not for you :(

awanlin replied:
I used this to help me understand the on-prem URLs: https://confluence.atlassian.com/confkb/the-differences-between-various-url-formats-for-a-confluence-page-278692715.html

But that was just to understand how it currently works and to understand the difference between the cloud URLs I was seeing. There does not seem to be docs for the Cloud URL like the above sadly

awanlin replied:
@adamdmharvey, any chance you'd be willing to log an issue about this and include some examples and ping me in it? That would be great as I'd think adding support for this should be straightforward but I'd like to keep this PR focused on the current changes

awanlin replied:
I've also added support for the URL format you shared, thanks 👍

tradcliffe-expediagroup commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/helpers.ts:
```diff
@@ -73,18 +76,20 @@ export const getAndWriteAttachments = async (
   const baseUrl = config.getString('confluence.baseUrl');
   const token = config.getString('confluence.token');
+  const isCloud = config.getOptionalBoolean('confluence.isCloud') || false;
+  const authToken = isCloud ? `Basic ${token}` : `Bearer ${token}`;
```
For this line, are you storing a base64 encoded username and password as your token variable. I wonder if instead we can update this to take in a username and password and encode it ? That way the user does not have to create the token locally, but we can create it for them? Then we can update the main functionality to do a check that either a token or username/password combo was supplied and if not throw an error.
Access to the non cloud rest api also supports basic auth and is something I was considering adding here soon. Which is why I am suggesting this.  Let me know what you think about this!

awanlin replied:
Yes, I took the path that Roadie.io used for their [Jira plugin](https://roadie.io/backstage/plugins/jira/#get-and-provide-jira_token-as-env-variable) where the user needs to do the encoding to avoid having more configuration (in our case we can now use the same token). I've got instructions on doing this in the README changes. For Confluence Cloud it's not a user name and password but your the email for your Atlassian account and the PAT like this: `<your-atlassian-account-mail>:<your-jira-token>`.

For this PR I'd like to stick with this and then maybe we can follow up with a PR focused on just this part?

tradcliffe-expediagroup replied:
Sure keep it like this for now. But it will need to be followed up on if there is a desire to allow non cloud users to log in with basic(logically.  A user could in theory set the isCloud to true even if it is not to use basic with an already encoded token!).  But if the precedent is that users do the encoding beforehand, and it is not done in the logic, we can keep the code using just a token and require encoding is done locally, like how you have it set up now.

awanlin replied:
I ended up doing a bit refactor and I've included this now, thanks 🚀

awanlin commented:
> I love this idea of adding more functionality to this. I added one comment as a suggestion/discussion.

Yeah, you've done some great work here and once I got thinks rolling I gave it some pretty messy, image and attachment heavy pages and it handle them like a boss! Awesome work @tradcliffe-expediagroup 

There are a few other changes I'd like to make but wanted to start with this so we could get using it. We actually don't have `mkdocs.yml` files in our repos - we were creating them and then we added support for creating a default one if there isn't #15417 - so it would be great to have an option to skip that part.

freben commented on plugins/scaffolder-backend-module-confluence-to-markdown/config.d.ts:
```diff
@@ -0,0 +1,34 @@
+    baseUrl: string;
+    /**
+     * The authentication token for accessing the Confluence API
+     * * @visibility secret
```
double asterisk here, i wonder if that actually would risk NOT marking it as secret 🤔 

I think it would make sense to document here too what the two supported token formats are, at least just very briefly.

freben commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/helpers.ts:
```diff
@@ -51,10 +51,13 @@ export const readFileAsString = async (fileDir: string) => {
   const token = config.getString('confluence.token');
-  const response: Response = await fetch(`${baseUrl}${relativeUrl}`, {
+  const isCloud = config.getOptionalBoolean('confluence.isCloud') || false;
+  const authToken = isCloud ? `Basic ${token}` : `Bearer ${token}`;
```
this whole dance seems to be suitable to break out into a separate function like

```ts
function parseConfig(config: Config): { authHeader: string; baseUrl: string }
```

or so, and use that below too

awanlin replied:
Yes, this got refactored into something I feel is much nicer 🚀

freben commented:
Just a small ping to keep it alive.

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/config.d.ts:
```diff
@@ -0,0 +1,34 @@
+    /**
+     * Flag to determine if you are on Confluence Cloud, optional
+     */
+    isCloud?: string;
```
Yeah, thanks for making this suggestions. I was worried about the dreaded [CLOUD-6999](https://jira.atlassian.com/browse/CLOUD-6999) but I ended up doing a big refactor and don't need this any more. It feels much cleaner.

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/confluenceToMarkdown.ts:
```diff
@@ -57,7 +57,7 @@ export const createConfluenceToMarkdownAction = (options: {
             title: 'Confluence URL',
             description:
-              'Paste your confluence url. Ensure it follows this format: https://{confluence+base+url}/display/{spacekey}/{page+title}',
+              'Paste your confluence url. Ensure it follows this format: https://{confluence+base+url}/display/{spacekey}/{page+title} or  https://{confluence+base+url}/spaces/{spacekey}/pages/1234567/{page+title} for Confluence Cloud',
```
Fixed 👍

awanlin commented:
@tradcliffe-expediagroup @freben I've done a large refactor of my changes based on feedback from both of you. Might be good to do another full review because of this. I added support for username/password auth and also moved the encoding into the action itself so people wanting to use it won't have to deal with that aspect.

awanlin commented:
After I did some manual testing I realized I inverted `basic` and `bearer`, that's been fixed now. Also, for the `userpass` auth method I pulled the implementation from here but I have no way to test it: https://github.com/backstage/backstage/blob/79ca6079f45e2765250d918ee50239568b490455/packages/integration/src/bitbucketCloud/core.ts#L130

freben commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/helpers.ts:
```diff
@@ -43,18 +43,93 @@ export interface Results {
+};
+
+export const getAuthorizationHeaderValue = (config: ConfluenceConfig) => {
+  let authHeaderValue: string = '';
```
i'd just delete this line and return in each case instead, no need for breaks then too :)

awanlin replied:
Sounds good, done 👍

freben reviewed (commented):
Alright, close!

freben reviewed (commented):
Alright, close!

adamdmharvey commented on .changeset/famous-days-kneel.md:
```diff
@@ -0,0 +1,5 @@
+'@backstage/plugin-scaffolder-backend-module-confluence-to-markdown': patch
+---
+
+Added support for Confluence Cloud to the `confluence:transform:markdown` action, view the [README](https://github.com/backstage/backstage/tree/master/plugins/scaffolder-backend-module-confluence-to-markdown) for more details
```
Based on the plug-in config , does this only work on Confluence Cloud? Those involved with Atlassian licensing may assume then that this action does not work with a Confluence Server, the on prem solution. Based on the baseUrls and such I feel like maybe it does work on both ? (Though I've not tested it!!)

awanlin replied:
I've updated the wording that this adds Confluence Cloud support along with the existing Confluence Server support.

adamdmharvey reviewed (commented):
I think this looks great. Minor comments. Thanks as always for great contribs to the community Andre!

awanlin commented on packages/backend/package.json:
```diff
@@ -60,6 +60,7 @@
     "@backstage/plugin-proxy-backend": "workspace:^",
     "@backstage/plugin-rollbar-backend": "workspace:^",
     "@backstage/plugin-scaffolder-backend": "workspace:^",
+    "@backstage/plugin-scaffolder-backend-module-confluence-to-markdown": "workspace:^",
```
Honestly I'm only adding it so that it's faster to get up and running for future work with this action. I'll try and make the changes you suggest

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/README.md:
````diff
@@ -56,15 +58,64 @@ export default async function createPlugin(
+```yaml
+confluence:
+  baseUrl: 'https://confluence.example.com'
+  auth: 'bearer'
````
Yeah, I've cleaned this up and it's more clear now. Thanks!

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/src/actions/confluence/confluenceToMarkdown.ts:
```diff
@@ -57,7 +58,7 @@ export const createConfluenceToMarkdownAction = (options: {
             title: 'Confluence URL',
             description:
-              'Paste your confluence url. Ensure it follows this format: https://{confluence+base+url}/display/{spacekey}/{page+title}',
+              'Paste your confluence url. Ensure it follows this format: https://{confluence+base+url}/display/{spacekey}/{page+title} or https://{confluence+base+url}/spaces/{spacekey}/pages/1234567/{page+title} for Confluence Cloud',
```
Fixed 👍

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/docs/sample-template.yaml:
Fixed 👍

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/docs/sample-template.yaml:
It is but in our internal version of this template we removed it as you needed to clear it out each time. I've done the same with this now.

awanlin commented on plugins/scaffolder-backend-module-confluence-to-markdown/config.d.ts:
```diff
@@ -0,0 +1,49 @@
+     */
+    token?: string;
+    /**
+     * Email encoded with the token for the bearer auth method
```
Yeah, this was for sure confusing. I've improved it based on your suggestion 👍

freben reviewed (approved):
Let's go!

awanlin commented:
Yeah!!! Thanks @freben 🚀
